roles/openshift_health_checker/library/search_journalctl.py

```diff
@@ -0,0 +1,155 @@
+    )
+
+    timestamp_limit_seconds = time() - 60 * 60  # 1 hour
+
+    log_count_limit = module.params["log_count_limit"]
+    log_matchers = module.params["log_matchers"]
+
+    matched_regexp, errors = get_log_matches(log_matchers, log_count_limit, timestamp_limit_seconds)
+
+    module.exit_json(
+        changed=False,
+        failed=len(errors),
+        errors=errors,
+        matched=matched_regexp,
+    )
+
+
+def get_log_matches(matchers, log_count_limit, timestamp_limit_seconds):
+    """Iterate through a list of matchers, verify required fields exist on each one, and
+    retrieve log output for each matcher, appending its match pattern to a collection if
+    a matcher's regex pattern is found.
+    Return a collection of matched regular expressions, and a list of errors."""
+    matched_regexp = []
+    errors = []
+
+    for matcher in matchers:
+        try:
+            ensure_matcher_fields(matcher)
+        except InvalidMatcherException as err:
+            errors.append(str(err))
+            continue
+
+        log_input, error = get_log_input(matcher)
+        if error:
+            errors.append(error)
+            continue
+
+        try:
+            matched, error = find_matches(log_input, matcher, log_count_limit, timestamp_limit_seconds)
+        except InvalidMatcherRegexpException as err:
+            matched = False
+            error = str(err)
+
+        if error:
+            errors.append(error)
+            continue
+
+        if matched:
+            matched_regexp.append(matcher["regexp"])
+
+    return matched_regexp, errors
+
+
+def ensure_matcher_fields(matcher):
+    """Ensure that expected fields exist on a user-defined matcher."""
+    required_fields = [
+        "start_regexp",
+        "regexp",
+        "unit",
+    ]
+    missing_fields = set(required_fields) - set([field for field in matcher])
+    if len(missing_fields):
+        raise InvalidMatcherException("Matcher missing required fields", list(missing_fields))
+
+
+def get_log_input(matcher):
+    """Run journalctl with the systemd unit specified in the matcher.
+    Return the command output as an iterator."""
+    try:
+        cmd_output = subprocess.Popen(list([
+            '/bin/journalctl',
+            '-ru', matcher["unit"],
+            '--output', 'json',
+        ]), stdout=subprocess.PIPE)
+
+        return iter(cmd_output.stdout.readline, ''), None
+
+    except subprocess.CalledProcessError as exc:
+        err = "Could not obtain journalctl logs for the specified systemd unit: {}: {}"
+        return None, err.format(matcher["unit"], str(exc))
+
+    except OSError as exc:
+        return None, str(exc)
+
+
+def find_matches(log_input, matcher, log_count_limit, timestamp_limit_seconds):
+    """Receive iterable input, a matcher dictionary and scan journalctl at the
+    given systemd module for log messages matching a given regular expression.
+    Return the matched log message (or None) and an error string if any."""
```

Comment:
It would be more natural if it just raise exceptions. Currently it communicates errors both by raising `InvalidMatcherRegexpException` AND return a string out of a `ValueError` it may catch...